### ISSUE TITLE
IPv6 + IPv4 switch to see if both are healty

### ISSUE THREAD

**Issue by resu-cireneg** [louislam/uptime-kuma#1242]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

API, New Monitor, UI Feature

### 🔖 Feature description

* currently it depends on whatever the environment is configured to prefer ( e.g. /etc/gai.conf )
* there is no proof if IPv4 + IPv6 Host return the same result

### ✔️ Solution

a little button "resolve ipv6+ipv6" or "dual-stack-check" and the respecting part in the monitoring stack

**Comment by BroOtti** [comment]
Seems to be a duplicate of #1025 
Also, it doesn't seem to depend on whatever the environment is configured to prefer. On default IPv6 is preferred, but uptime kuma still checks on IPv4 and not on IPv6. Properly, I could just disable IPv4 on my uptime kuma host, but that way I need to activate it again if I want to update uptime kuma as GitHub only supports IPv4... 
The best solution would be to add the option of which IP to use, but at least uptime kuma should follow the standard and use IPv6 as default.

**Comment by JulianGro** [comment]
The problem for OP and me is that we cannot check that IPv4 *and* IPv6 works.
#1025 isn't strictly the same because that is about switching between IPv4 and IPv6 instead of testing both.

**Comment by Henk57** [comment]
Same issue with me. On a location a NAS is only on IPv6.  Uptime kuma sees this server as down while it is up and running with a valid DNS and FQDN.

**Comment by dtrckd** [comment]
see how https://updown.io/ handle ipv4 + ipv6, not even need to switch, it would be  a really nice feature.

**Comment by RaitoBezarius** [comment]
Am interested into implementing this as I am running into a lot of false positives due to IPv4/IPv6 thing.

**Comment by traylenator** [comment]
Even a tick box for ipv4 or ipv6 only and then easy enough to create two monitors if needs be - probably easier.

**Comment by ProjectExy** [comment]
> Even a tick box for ipv4 or ipv6 only and then easy enough to create two monitors if needs be - probably easier.

Im hoping for this

**Comment by larsch** [comment]
Generally speaking, a single URL with hostname can resolve to both IPv4 and IPv6 address(es), but also MULTIPLE addresses. An uptime monitor should really be checking all resolutions of the domain name, not just the default or first one it happens to get.

**Comment by stefanux** [comment]
> An uptime monitor should really be checking all resolutions of the domain name, not just the default or first one it happens to get.

not sure if there is any monitoring solution which checks both A/AAAA and possibly even multiple records of them. seems out of scope for uptime kuma since you could simply add more checks for everything you need (v4 and v6).

if you have multiple A/AAAA-Records for http you normally would check on IP and use specific hostname to get correct site, this could be done via 
setting "URL" to "https://1.2.3.4/my-monitoring-uri" and send host header using "JSON-Headers":

```
{
    "Host": "www.MY.HOST.TLD"
}
```

most basic is the feature to choose address family for all protocols when using DNS.

**Comment by CommanderStorm** [comment]
> if you have multiple A/AAAA-Records for http you normally would check on IP and use specific hostname to get correct site, this could be done via

Please not this bug in ipv6 handling of axios: https://github.com/louislam/uptime-kuma/issues/3564
Fixing said bug would require an upgrade of axios which is blocked by there not being a documentation on how to upgrade from 0.x to 1.x and what changes were made.

**Comment by larsch** [comment]
> not sure if there is any monitoring solution which checks both A/AAAA and possibly even multiple records of them. seems out of scope for uptime kuma since you could simply add more checks for everything you need (v4 and v6).

You could. I still think the current model can be dangerous for the naive user. You get a green label, even if your connectivity is partially broken. I'm not complaining here, just suggesting that an implicit check of all domain name resolutions would what the user really needs, and often didn't realise that they wanted.

> if you have multiple A/AAAA-Records for http you normally would check on IP and use specific hostname to get correct site, this could be done via setting "URL" to "https://1.2.3.4/my-monitoring-uri" and send host header using "JSON-Headers":
> 
> ```
> {
>     "Host": "www.MY.HOST.TLD"
> }
> ```

You would also need to be able to specify SNI for this work.

**Comment by lucasnz** [comment]
I've been working on a PR to resolve this. It would add an option to specify IPv4 or IPv6. If you leave it on "auto detect" (the default) it will use the current logic. It works with the following monitors so far: Ping, TCP, DNS. I'm working on the HTTP monitor then I'll submit it.

Here are some screenshots of the option:
![Screenshot 2023-10-12 174951](https://github.com/louislam/uptime-kuma/assets/606314/d9a6cae8-058d-4873-83ff-0fe2770fb83e)
![Screenshot 2023-10-12 175028](https://github.com/louislam/uptime-kuma/assets/606314/1f88e4fa-ebfb-475b-a9fd-111d9926334c)

**Comment by MatzeKitt** [comment]
Thank you!

I’m not sure if "IP Family" is the correct term here, as to my knowledge, it refers to a broader set of IP protocols as TCP, DNS, SMTP etc. – all being part of the IP family. Thus I would suggest going with "IP version".

**Comment by lucasnz** [comment]
I'm not particularly set on the name. It's called IP Address Family in normally, e.g. in the nodejs doco it calls it address family: https://nodejs.org/api/net.html#socketaddressfamily

**Comment by ev0lister** [comment]
IP address family should be fine to use, as per [IANA](https://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml) the family version 4 and 6 have assigned number 1 and 2.

That being said, I appreciate the work to be able to determine the ip protocol being used for a monitor.

**Comment by Derkades** [comment]
Thank you for your work on this! It would be great if there was a fourth option to check both IPv4 and IPv6, avoiding the need to create 2 ping jobs. Not sure if that's possible in the current uptime-kuma framework.

**Comment by lucasnz** [comment]
> Thank you for your work on this! It would be great if there was a fourth option to check both IPv4 and IPv6, avoiding the need to create 2 ping jobs. Not sure if that's possible in the current uptime-kuma framework.

I don't see how you would implement a dual check. If IPv4 was down, but IPv6 was up - how would the logic of the monitor be set - should it show as down or up. How would it display which IP version wasn't working? I think it best that this is set up as two monitors for clarity.

**Comment by joachimtingvold** [comment]
If a service is supposed to be dual-stacked, but suddenly becomes single-stacked (regardless of it's IPv4-only or IPv6-only), the service is for all intents and purposes "down" the way I see it. If Kuma monitors both IPv4 and IPv6 simultaneously, it's no problem to alert/identify what address-family that becomes unresponsive (and alert accordingly, like "IPv4 down for service XYZ" or similar). If the "down" vs. "partially-down" is a controversal topic, it could probably be made as an option (i.e. if it should just alert, but keep the service status "up"), or maybe even better; introduce "Degraded" status (in addition to "Up" and "Down"). The latter could probably also be tied in with other feature requests (like #2313).

**Comment by stefanux** [comment]
@joachimtingvold you cant assume clients are dual-stacked and perform some kind downgrade to v4.
v6 only (or v4 only) cant reach your services therefore its down for them (not a service "degradation" which would imply reachable but slower).

**Comment by joachimtingvold** [comment]
That's essentially what I said; personally I consider a dual-stacked service to be down if it for some reason becomes single-stacked (regardless if IPv4 or IPv6 is broken). However, I merely suggested that it could be a configuration choice. As an example, I assume some use Kuma for internal services, and if you have control of the environment, there might be scenarios where you would like to consider the scenario of "dual-stacked service that becomes single-stacked" to be "Degraded", rather than "Down".

**Comment by joachimtingvold** [comment]
I still think it would be beneficial to have options, and let users decice how they want to monitor things. As I mentioned, I see usecases for both variants in controlled environments. If this is implemented as a IPv4/IPv6-dualstacked-specific feature, or combined with the "service grouping"-feature that has been implemented (or at least partially, see #2487 + #2474 + #2693 + #1236, and probably others), I have no strong opinion on. There are many similar problems that dual-stacked monitoring and service grouping have in common. Some of the top of my head;

- Mark the entire service as down if IPv4 or IPv6 is down
- Mark service as degraded if IPv4 or IPv6 is down
- Alert if one part of the service is down (I would like to know specifically if IPv4 or IPv6 failed)
- Don't alert for both IPv4 and IPv6 if the entire service goes down (just alert once that the service is down)
- ... and probably more

**Comment by stefanux** [comment]
most tools i know have explicit -4 or -6 switches, or (by default) do downgrades by default (for which you might want to set the order if v6 > v4 is not desired). not a fan of introducing another status (aside up or down) which would need extra settings in notifications.

**Comment by troykelly** [comment]
There's already extensive support for this clearly, adding my voice to that.

I'd also like to share my thinking on the handling of a v4 and/or v6 outage.

IPv4 and IPv6 are entirely different networks, with different routing and, different network-level issues that could impact performance.

Yes, they are typically handled by the same server, often the same service - but that's not a requirement.

There are IPv4-only networks, dual-stack and IPv6 only networks.

In my opinion, the ability to select the IP family needs to be thought of simply as a time saver for creating two entries - and that's it.

An outage of IPv4 or IPv6 while the other remains up is _not_ a degradation of service - it's an outage - the server is offline for some (or maybe all) of your eyeballs.

The failure of _both_ IPv4 and IPv6 may be treated differently by some - perhaps it means some more significant outage of the service or backend rather than a network layer issue - but that really depends on the service, the environment and those who manage it.

IPv6 is a replacement for IPv4 (eventually... maybe... someday?) - we must not make the mistake of thinking that _dual stack for life_ is the solution, and our monitoring tools need to embrace that thinking also.

**Comment by ShadowJonathan** [comment]
Not even `gai.conf` seems to properly order what is being preferred, since axios seems to pick ipv6 over ipv4: https://github.com/axios/axios/issues/5785

**Comment by Crash1602** [comment]
I have the same issue. Some systems are only accessible via IPv6, but my server defaults resolve to IPv4. The ability to monitor both or choose whether to resolve IPv4 or IPv6 would be a very helpful feature.

**Comment by joachimtingvold** [comment]
> In my opinion, the ability to select the IP family needs to be thought of simply as a time saver for creating two entries - and that's it.
> […] 
> Where we can, we already create separate monitors for each IP family:

It's not just "to save time" (by not having to create duplicate rules).

99% of the services I'd like to monitor is using FQDNs with A+AAAA records. I don't want to create A- and AAAA-only FQDNs for the same service (ipv4.service.foobar.com + ipv6.service.foobar.com, etc), just for the sake of being able to monitor the IP-stacks individually.

I'm not interested in using their IP addresses either (i.e. for services that are using dynamic DNS entries or similar). To facilitate this, as a minimum you'd need some kind of "address family" option for the alert (where you can choose between IPv4 or IPv6).

As a "phase 2" of the address family option, you could implement full dual-stack alerting/monitoring for a rule, and uptime-kuma can figure out this by itself (by looking at the DNS response), and monitor one or both of the address families for a given FQDN.

**Comment by Aokromes** [comment]
Lately on spain laliga is blocking some hosts using ipv4 blocks, if you have ipv6 you can connect to them that's why allow checking domains on both ipv4 and ipv6 is wanted.

**Comment by troykelly** [comment]
To hopefully add a little more clarity, here's an example if I were monitoring the Apple website.

If I'm testing `www.apple.com` I need to find _every_ server hosting it (or load balancing, or reverse proxying, or the other things). I then need to test every single one of them. If _everything_ is available - the test is a success.

This type of test is nearly impossible with Uptime Kuma - it (at the very least) requires the configuration of separate tests for each load balancer, hard coding of IP's etc. This doesn't work in a dynamic environment at all.

The script is available here: https://gist.github.com/troykelly/0d87ef3add9f0767437303343e075ea8#file-check-lb-py

This is _only_ an example for web, other dual-stack services DNS, voip, etc all need their own handling of IPv4 and IPv6.

<img width="1361" alt="Image" src="https://github.com/user-attachments/assets/a5acdb8a-4715-4599-8463-db085acc4447" />

```zsh
root@docker-desktop:/app# ./check-lb.py https://www.apple.com/ --timing
Load Balancer Check for domain: www.apple.com
Original URL: https://www.apple.com/
DNS | Hostname                                                                      | IP                       | Proto | St     | ConnT | RespT | Redirect / Error      
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
🌐   | a23-52-247-118.deploy.static.akamaitechnologies.com                           | 23.52.247.118            | http  | 301 ↪️ | 2.3   | 3.1   | https://www.apple.com/
🌐   | g2600-1415-9c00-018e-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com | 2600:1415:9c00:18e::1aca | http  | 301 ↪️ | 2.5   | 2.1   | https://www.apple.com/
🌐   | g2600-1415-9c00-019d-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com | 2600:1415:9c00:19d::1aca | http  | 301 ↪️ | 2.5   | 2.2   | https://www.apple.com/
🌐   | a23-52-247-118.deploy.static.akamaitechnologies.com                           | 23.52.247.118            | https | 200 ✅ | 39.9  | 16.3  |                       
🌐   | g2600-1415-9c00-018e-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com | 2600:1415:9c00:18e::1aca | https | 200 ✅ | 36.5  | 8.8   |                       
🌐   | g2600-1415-9c00-019d-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com | 2600:1415:9c00:19d::1aca | https | 200 ✅ | 38.0  | 7.3   |                       

Overall LB check: SUCCESS
root@docker-desktop:/app# ./check-lb.py https://www.apple.com/ --timing --json
{
  "domain": "www.apple.com",
  "original_url": "https://www.apple.com/",
  "overall_status": "success",
  "results": [
    {
      "dns_source": "A/AAAA",
      "ip": "23.52.247.118",
      "hostname": "a23-52-247-118.deploy.static.akamaitechnologies.com",
      "family": "IPv4",
      "scheme": "http",
      "status": 301,
      "location_header": "https://www.apple.com/",
      "error_message": null,
      "connect_time_ms": 1.681,
      "response_time_ms": 2.887
    },
    {
      "dns_source": "A/AAAA",
      "ip": "2600:1415:9c00:18e::1aca",
      "hostname": "g2600-1415-9c00-018e-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com",
      "family": "IPv6",
      "scheme": "http",
      "status": 301,
      "location_header": "https://www.apple.com/",
      "error_message": null,
      "connect_time_ms": 2.083,
      "response_time_ms": 1.978
    },
    {
      "dns_source": "A/AAAA",
      "ip": "2600:1415:9c00:19d::1aca",
      "hostname": "g2600-1415-9c00-019d-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com",
      "family": "IPv6",
      "scheme": "http",
      "status": 301,
      "location_header": "https://www.apple.com/",
      "error_message": null,
      "connect_time_ms": 1.667,
      "response_time_ms": 1.924
    },
    {
      "dns_source": "A/AAAA",
      "ip": "23.52.247.118",
      "hostname": "a23-52-247-118.deploy.static.akamaitechnologies.com",
      "family": "IPv4",
      "scheme": "https",
      "status": 200,
      "location_header": null,
      "error_message": null,
      "connect_time_ms": 43.267,
      "response_time_ms": 28.482
    },
    {
      "dns_source": "A/AAAA",
      "ip": "2600:1415:9c00:18e::1aca",
      "hostname": "g2600-1415-9c00-018e-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com",
      "family": "IPv6",
      "scheme": "https",
      "status": 200,
      "location_header": null,
      "error_message": null,
      "connect_time_ms": 37.553,
      "response_time_ms": 7.246
    },
    {
      "dns_source": "A/AAAA",
      "ip": "2600:1415:9c00:19d::1aca",
      "hostname": "g2600-1415-9c00-019d-0000-0000-0000-1aca.deploy.static.akamaitechnologies.com",
      "family": "IPv6",
      "scheme": "https",
      "status": 200,
      "location_header": null,
      "error_message": null,
      "connect_time_ms": 37.404,
      "response_time_ms": 7.512
    }
  ]
}
```
